### PR TITLE
JDK-6229853: BasicTextAreaUI:create incompletely documents the possible returned View types

### PR DESCRIPTION
**Pull request by prsadhuk**
API specification says about BasicTextAreaUI.create "Returns a WrappedPlainView or PlainView". 
But for [bidi](https://github.com/openjdk/jdk/blob/d14e96d9701dae951aa365029f58afb6687a646a/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java#L168) text, this method returns BoxView and other views , so the documentation needs to be updated.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-6229853](https://bugs.openjdk.org/browse/JDK-6229853): BasicTextAreaUI:create incompletely documents the possible returned View types
 * [JDK-8294114](https://bugs.openjdk.org/browse/JDK-8294114): BasicTextAreaUI:create incompletely documents the possible returned View types (**CSR**)

### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10376/head:pull/10376` \
`$ git checkout pull/10376`

Update a local copy of the PR: \
`$ git checkout pull/10376` \
`$ git pull https://git.openjdk.org/jdk pull/10376/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10376`

View PR using the GUI difftool: \
`$ git pr show -t 10376`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10376.diff">https://git.openjdk.org/jdk/pull/10376.diff</a>

</details>

### CODE REVIEW

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java [hunk 1]
```diff
@@ -138,7 +138,8 @@ public Dimension getMinimumSize(JComponent c) {
 
     /**
      * Creates the view for an element.  Returns a WrappedPlainView or
-     * PlainView.
+     * PlainView if non-bidi text is handled
+     * else returns PlainParagraph or BoxView if bidi text is being handled.
      *
```
Comment (prrace) [inline src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java:143]: So .. what if the text wasn't bidi but then some one adds bidi text ?

And can we change the summary of the BUG and PR to something grammatically correct, and perhaps even
descriptive such as
"BasicTextAreaUI:create incompletely documents the possible returned View types"

Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java:143]: In that case BasicTextUI#insertUpdate() will change the model which will again call BasicTextAreaUI#create to change the view to i18n to either of BoxView or PlainParagraph.

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java [hunk 2]
```diff
@@ -138,7 +138,8 @@ public Dimension getMinimumSize(JComponent c) {
 
     /**
      * Creates the view for an element.  Returns a WrappedPlainView or
-     * PlainView.
+     * PlainView if non-bidi text is handled
+     * else returns PlainParagraph or BoxView if bidi text is being handled.
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java:142]: You cannot specify that this method returns `PlainParagraph`, it's a private class. I think we should relax the specification to allow other types to be returned without mentioning the exact types.

This issue is also relevant to `BasicTextFieldUI`, yet [it does not state explicitly](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/plaf/basic/BasicTextFieldUI.html#create(javax.swing.text.Element)) what types are valid, it only gives a hint that a `FieldView` is returned:

> Creates a view (FieldView) based on an element.

So, I propose the following text: *“Creates a view for an element. Usually returns a WrappedPlainView or PlainView.”*

The `@return` tag may be changed to “a view (for the the element)” for this class as well as for `BasicTextFieldUI`.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

**Comment by prrace** [comment]
Yeah I don't  like the way this is documented at all. And it seems this can be swapped out on the fly when content changes too.
I think we need to start with WHY we need to document the return types. And if we don't then we don't.
If we do I'd like some detailed reasoning and contemplation of what we say about if it changes later.

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java (hunk 2 above)
Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java:142]: I guess we can mention "Creates a view for an element." only.
`Usually returns a WrappedPlainView.. ` can give problem to TCK team to create testcase for all scenario. Either we mention the full `returns WrappedView or PlainView. or ParagraphView or BoxView`

**Comment by aivanov-jdk** [comment]
> Yeah I don't like the way this is documented at all. And it seems this can be swapped out on the fly when content changes too.

`BasicTestUI` implements `ViewFactory` which supplies views for text components.

The type of view returned depends on the document content, in particular whether it contains bidirectional text or not.

> I think we need to start with WHY we need to document the return types. And if we don't then we don't. If we do I'd like some detailed reasoning and contemplation of what we say about if it changes later.

We don't need to. The views for regular text are in public API: [`FieldView`](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/text/FieldView.html) that's used by `BasicTextFieldUI` as well as [`PlainView`](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/text/PlainView.html) and [`WrappedPlainView`](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/text/WrappedPlainView.html) which are used by `BasicTextAreaUI`. Yet the views for bidirectional text aren't public. So it looks, the spec hints at the public types of views returned for the regular text, yet it says nothing about package-private views which support bidirectional text.

`BasicTextFieldUI` says the view is likely `FieldView` in parentheses, but `BasicTextAreaUI` states it returns `WrappedPlainView` or `PlainView` (only), which is incorrect.

Can we leave the reference to public API classes?  
Should we remove the references to types completely? If yes, `BasicTextFieldUI.create` should be updated as well.

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java (hunk 2 above)
Comment (aivanov-jdk) [reply src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java]: > I guess we can mention "Creates a view for an element." only. `Usually returns a WrappedPlainView.. ` can give problem to TCK team to create testcase for all scenario.

Then the current code shouldn't have passed the TCK as the spec requires either `WrappedView` or `PlainView`.

> Either we mention the full `returns WrappedView or PlainView. or ParagraphView or BoxView`

We can't specify it returns a package-private class.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java [hunk 3]
```diff
@@ -137,12 +137,10 @@ public Dimension getMinimumSize(JComponent c) {
     }
 
     /**
-     * Creates the view for an element.  Returns a WrappedPlainView or
-     * PlainView if non-bidi text is handled
-     * else returns PlainParagraph or BoxView if bidi text is being handled.
+     * Creates the view for an element.
```
[the remark this answers is not in the document]
Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextAreaUI.java]: ok

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextFieldUI.java [hunk 4]
```diff
@@ -86,7 +86,7 @@ protected String getPropertyPrefix() {
      * Creates a view (FieldView) based on an element.
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextFieldUI.java:86]:
```suggestion
     * Creates a view based on an element.
```
For consistency, if we don't document the types.

Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextFieldUI.java]: ok

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextFieldUI.java [hunk 5]
```diff
@@ -83,10 +83,10 @@ protected String getPropertyPrefix() {
     }
 
     /**
-     * Creates a view (FieldView) based on an element.
+     * Creates a view based on an element.
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextFieldUI.java:86]: Should we use the wording from `BasicTextUI` for consistency?

```suggestion
     * Creates a view for an element.
```

The description in [`ViewFactory.create`](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/text/ViewFactory.html#create(javax.swing.text.Element)) is more detailed, a short sentence as above is enough for `TextUI`.

**Review by aivanov-jdk — APPROVED** [review]
Looks good if our decision is not document the types of views returned.

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextFieldUI.java (hunk 5 above)
Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTextFieldUI.java]: Seems ok for consistency..
Have updated the PR and CSR...Can you please review CSR https://bugs.openjdk.org/browse/JDK-8294114

**Review by aivanov-jdk — APPROVED** [review]

**Review by prrace — APPROVED** [review]

**Pull request closed without merging**